test/mpi/errors/spawn/connect_timeout.c

```diff
@@ -118,11 +126,11 @@ int test_mismatched_accept(MPI_Comm intra_comm, int gid)
         /* NOTE: if accept hangs, try increase MPIR_CVAR_CH3_COMM_CONN_TIMEOUT. */
         IF_VERBOSE(intra_rank == 0, ("server: accepting connection to <%s>.\n", port));
         MPI_Comm_accept(port, MPI_INFO_NULL, 0, intra_comm, &comm);
-    }
+        if (comm != MPI_COMM_NULL) {
+            IF_VERBOSE(intra_rank == 0, ("connection matched, freeing communicator.\n"));
+            MPI_Comm_free(&comm);
+        }
 
-    if (comm != MPI_COMM_NULL) {
-        IF_VERBOSE(intra_rank == 0, ("connection matched, freeing communicator.\n"));
-        MPI_Comm_free(&comm);
     }
```

Comment:
I am a bit confused about this change. It looks like you just move the `if (comm != MPI_COMM_NULL){}` block into the above `if---else if---`. Why do we do it?

Reply:
There are two `MPI_Comm_connect`, one of them will succeed and the other one will return `MPI_COMM_NULL`. If the first one succeed, the second one (failed ones) will set comm to `MPI_COMM_NULL` and the comm objects from the first successful one will get leaked.

Since we don't know which `MPI_Comm_connect` will succeed, we need test each and free the one that succeeds (when `comm != MPI_COMM_NULL`).